modules/core/src/directx.cpp

```diff
@@ -1392,11 +1393,201 @@ void convertFromD3D11Texture2D(ID3D11Texture2D* pD3D11Texture2D, OutputArray dst
+        if (clGetDeviceIDsFromD3D11KHR)
+        {
+            // try with CL_PREFERRED_DEVICES_FOR_D3D11_KHR
+            do {
+                status = findKHRDevice(clGetDeviceIDsFromD3D11KHR,
+                                       pD3D11Device,
+                                       CL_PREFERRED_DEVICES_FOR_D3D11_KHR,
+                                       platforms[i],
+                                       devices,
+                                       found_preffered);
+                if (status != CL_SUCCESS) break;
+            } while (0);
+
+            // try with CL_ALL_DEVICES_FOR_D3D11_KHR
+            if (!found_preffered) do {
+                status = findKHRDevice(clGetDeviceIDsFromD3D11KHR,
+                                       pD3D11Device,
+                                       CL_ALL_DEVICES_FOR_D3D11_KHR,
+                                       platforms[i],
+                                       devices,
+                                       found_preffered);
+                if (status != CL_SUCCESS) break;
+            } while (0);
+        }
+    }
+#endif
+}
+
+std::tuple<cv::ocl::Image2D, cv::ocl::Context> convertFromD3D11Texture2DtoCLMem(ID3D11Texture2D* pD3D11Texture2D, Device& device)
+{
+    CV_UNUSED(pD3D11Texture2D); CV_UNUSED(device);
+#if !defined(HAVE_DIRECTX)
+    NO_DIRECTX_SUPPORT_ERROR;
+#elif !defined(HAVE_OPENCL)
+    NO_OPENCL_SUPPORT_ERROR;
+#else
+    if (device.ptr() == nullptr) {
+        CV_Error(cv::Error::StsNullPtr, "OpenCL: convertFromD3D11Texture2DtoCLMem got empty Device object");
+    }
+
+    cl_device_id cl_device = (cl_device_id)device.ptr();
+    cl_platform_id platform = nullptr;
+
+    size_t out_size = 0;
+    cl_int status = clGetDeviceInfo(cl_device, CL_DEVICE_PLATFORM, 0, nullptr, &out_size);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clGetDeviceInfo failed");
+    }
+    status = clGetDeviceInfo(cl_device, CL_DEVICE_PLATFORM, out_size, &platform, nullptr);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clGetDeviceInfo failed");
+    }
+    if (platform == nullptr) {
+        CV_Error(cv::Error::StsNullPtr, "OpenCL: clGetDeviceInfo returns empty platform");
+    }
+    auto&& platformName = PlatformInfo(&platform).name();
+
+    ID3D11Device *pD3D11Device = nullptr;
+    pD3D11Texture2D->GetDevice(&pD3D11Device);
+
+    if (pD3D11Device == nullptr) {
+        CV_Error(cv::Error::StsNullPtr, "OpenCL: GetDevice returns empty pD3D11Device in convertFromD3D11Texture2DtoCLMem");
+    }
+
+    cl_context context = NULL;
+    cl_context_properties properties[] = {
+            CL_CONTEXT_PLATFORM, (cl_context_properties)platform,
+            CL_CONTEXT_D3D11_DEVICE_KHR, (cl_context_properties)(pD3D11Device),
+            CL_CONTEXT_INTEROP_USER_SYNC, CL_FALSE,
+            NULL, NULL
+    };
+    context = clCreateContext(properties, 1, &cl_device, NULL, NULL, &status);
+    if (status != CL_SUCCESS && cl_device != nullptr)
```

Comment:
cannot understand this condition..
`cl_device` must be available at ` cl_device_id cl_device = (cl_device_id)device.ptr();` step - thus do not need to check it here

Since you use cl_device as inner resource of `Device& device` then you must not release it here in comparison with latter version of code where `cl_device` had created in-place. We should avoid release resources which we did not acquire (in comparison with context here)

Also after this function failed - you should not proceed here, because if we cannot create context then nothing else matter. Also i suggest to throw exception here and after in case of bad openCL statuses

Reply:
Added CV_Error